ENH add ratio-of-uniforms method for rv generation to scipy.stats

chrisb83 opened the pull request:
General method to sample random variates based on the density only. 

- can be useful to generate rvs in case a distribution is not part of scipy.stats and cdf cannot be easily inverted (such as the generalized inverse Gaussian). 
- I intend to add the generalized inverse Gaussian in a separate project based on this method (approach is referenced in the docstring which should also fix current restriction on parameters of `invgauss.rvs` as a special case)
- once gen. inverse Gaussian is available, one can also add generalized hyperbolic distributions

Tests to be added, but I checked that the approach works in general using `stats.kstest` and a couple of examples (see docstring).

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,170 @@ def rankdata(a, method='average'):
+        ensures successful simulation with a high probability.
+    bounds : tuple of floats (a, b), optional
+        Default is none. In case any of the boundaries of the bounding
+        rectangle is not specified, attempt to find missing boundary
```
is -> are, I think

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,170 @@ def rankdata(a, method='average'):
+    If the bounding rectangle is not correctly specified (i.e. if it does not
+    contain A(c)), the algorithm samples from a distribution different from
+    the one given by the density f. It is therefore recommended to perform a
+    test such as `stats.kstest` as a check.
```
Any comment on the time / space complexity of this algorithm -- especially given that there are academic citations below?

chrisb83 replied:
I added a comment on expected number of iterations.

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,170 @@ def rankdata(a, method='average'):
+    - x2 = sup (x - c) sqrt(f(x))
+    - y2 = sup sqrt(f(x))
+
+    In particular, these values are finite if f is bounded and x**2 f(x) is
```
Is there supposed to be an operator between x**2 and f(x) here?

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,170 @@ def rankdata(a, method='average'):
+        return opt_res.fun
+
+    # in case bounding rectangle is not given, attempt to find it numerically
+    def f_opt(x): return (x - shift) * np.sqrt(f(x))
```
I think we can probably afford to put the return statement on the next line here. I wonder if defining _private_functions() in the same module and using those within `ratio_unif` is an alternative option worth considering? I guess I don't have a strong reason for suggesting that apart from perhaps potential clarity from avoiding nested function definitions.

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,170 @@ def rankdata(a, method='average'):
+    # start sampling using ratio of uniforms method
+    x = np.zeros(size)
+    j, simulated = 1, 0
+    while (j <= maxiter):
```
May be some potential for Cythonization in the future here?

andyfaff replied:
I'd prefer a for loop here:
```
for j in range(maxiter):
```

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,170 @@ def rankdata(a, method='average'):
+            x[simulated:(simulated + take)] = rvs[accept][0:take]
+            simulated += take
+        if simulated < size:
+            j = j + 1
```
Is there a compelling reason to use the long-form increment of `j` here while the augmented operator is being used for the `simulated` variable just two lines above?

chrisb83 commented:
@tylerjereddy : thanks for your comments! I changed the code accordingly.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
+    """
+    Generate random variables using ratio-of-uniforms method.
+
+    Given a continuous univariate density f and a constant c, define the set
```
If this is going to be part of the public API then I think the language needs simplifying significantly.
Does `f` represent the PDF? I would use PDF instead because all the stats distributions have a PDF method.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
+    Generate random variables using ratio-of-uniforms method.
+
+    Given a continuous univariate density f and a constant c, define the set
+    A(c) := {(x, y) : 0 < y <= sqrt(f(x/y + c)) }. If (U, V) is a random
```
This mathematical notation won't be easy to understand by a lot of people. If it's not necessary to know this to use the function then I would put it in the notes.

chrisb83 replied:
i tried to simplify it, but the definition of the set A(c) is crucial for understanding the algorithm

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
+def ratio_unif(f, size=1, x1=None, x2=None, y2=None, shift=0, maxiter=None,
+               bounds=None):
+    """
+    Generate random variables using ratio-of-uniforms method.
```
Generate random samples from a PDF using the ratio-of-uniforms method.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
+    with density f.
+
+    The above result (see [1]_, [2]_) can be used to sample random variables
+    using only the density. Typical choices of c are zero or the mode of f.
```
Might want to outline the circumstances for when it's going to be better/faster to use this function than calculate via pdf-->cdf-->ppf-->rvs.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
+    number of iterations to draw (U, V) uniformly distributed on R(c) such that
+    (U, V) is also in A(c) is given by the ratio
+    area(R(c)) / area(A(c)) = 2 * y2 * (x2 - x1), using the fact that
+    the area of A(c) is equal to 1/2 (Theorem 7.1 in [1]_).
```
I would relegate 5730 to 5752 to the notes. Overly mathematical for the initial part of the docstring.

chrisb83 replied:
I moved some more material to the notes, however, some parts need to stay to understand the method in my view

andyfaff replied:
The text at the start of the function does need to be kept, but in this notes section.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
+    Parameters
+    ----------
+    f : callable
+        Scalar function defining the density.
```
Is f the PDF? If so, then use PDF instead. What is meant by a 'scalar function'? Do you mean it takes a scalar, or that the output is a scalar? Can this function deal with `f` that can be vectorised?

chrisb83 replied:
changed, it is simply the PDF, i dropped the term scalar which is confusing here

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
+    ----------
+    f : callable
+        Scalar function defining the density.
+    size : int, optional
```
The normal behaviour for size in scipy.stats is:

> size : int or tuple of ints, optional
    Defining number of random variates (default is 1).

i.e. you can either ask for a single value, a vector, or an N-D array of samples (by giving a shape)

chrisb83 replied:
ok, i haven't changed this yet, but should not be difficult.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
+        Default is None. In that case, attempt to find bound numerically.
+    y2 : float, optional
+        The upper bound of the bounding rectangle in the y-direction.
+        Default is None. In that case, attempt to find bound numerically.
```
what are the tolerances on finding the lower and upper bounds? How close to the tails does one go?

andyfaff replied:
Should the tolerances be specifiable?

chrisb83 replied:
you mean tolerance for the optimize.minimize_scalar? I would just go with the default setting

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
     return .5 * (count[dense] + count[dense - 1] + 1)
+
+
+def ratio_unif(f, size=1, x1=None, x2=None, y2=None, shift=0, maxiter=None,
```
Perhaps a better function name is `rvs_ratio_unif`. `rvs` indicates that random variates are involved.

chrisb83 replied:
agreed

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
+        Maximum number of iterations to generate the random variables.
+        Default is None. In that case, maxiter is set to a value that
+        ensures successful simulation with a high probability.
+    bounds : tuple of floats (a, b), optional
```
It's not clear what `bounds` refers to. Does this mean bounds in `x` or in `y`? Or does it refer to vertices?

andyfaff replied:
How is this argument different to x1, x2, y2? Where is y1, is it taken to be equal to y2?

chrisb83 replied:
I changed the docstring, hopefully it is more clear now. y1 is always zero, I used y2 to indicate it is also the upper bound of the rectable like x2. Maybe there is a better choice...

andyfaff replied:
Is it necessary to specify x1/x2 *and* bounds?

chrisb83 replied:
no, bounds are only required if any of the Parameters x1, x2, y2 is not specified. clarified it in the doc string

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
+    >>> rvs = ratio_unif(f, x1=x1, x2=x2, y2=y2, size=2500)
+
+    The K-S test confirms that the random variates are indeed normally
+    distributed (p-value clearly does not reject normality):
```
p-value does not reject normality

chrisb83 commented:
@andyfaff thanks for your review as well! I tried to address most of your comments in my last commit and tried to simplify the description. I agree that understanding the method is not trivial, but i would say it is also not too hard (in the end just sampling uniform rvs and checking if their ratio lies in a certain set). it is an advanced method that hopefully has a clear use case when the standard approch via inversion of the cdf fails (like in #2648, solved in #6739, which still needs to be reviewed by the way :) )To add distributions like generalized inverse Gaussian / hyperbolic distributions, this seems to be one of the best methods, see the second ref in the docstring.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
     return .5 * (count[dense] + count[dense - 1] + 1)
+
+
+def rvs_ratio_unif(f, size=1, x1=None, x2=None, y2=None, shift=0, maxiter=None,
```
Coming from an unrelated area I think it would work better if it was:
`rvs_ratio_unif(pdf, ...`

andyfaff replied:
`shift` isn't described well in the list of parameters.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+    Generate random samples from a probability density function using the
+    ratio-of-uniforms method.
+
+    Given a univariate probability density function (pdf) f and a constant c,
```
Please put 5731 to 5746 in the notes section. That's normal for something like this. I'm not requesting that you remove it completely, it just doesn't belong here.
For example [`stats.norm`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.norm.html#scipy.stats.norm) describes it's PDF in the notes.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+
+    Parameters
+    ----------
+    f : callable
```
```
pdf: callable
    The probability density function of the distribution

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+    f : callable
+        The pdf of the distribution.
+    size : int, optional
+        The number of random variables to generate.
```
```
size : int or tuple of ints, optional

Defining number of random variates (default is 1).
```
Tuple of ints should generate an ND array of random variates.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+    Returns
+    -------
+    rvs : array
+        array with the random variables distributed according to pdf f
```
the random variables distributed according to the probability distribution, `pdf`

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+
+    """
+
+    if not callable(f):
```
I wouldn't test for callability here, [EAFP](https://docs.python.org/3/glossary.html#term-eafp)

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+    if x2 is None:
+        x2 = -find_rect_bounds(lambda x: -f_opt(x), bounds=bounds)
+
+    if x1 >= x2:
```
x1/x2 just bookend a range. Why not just do:
```
x1, x2 = min(x1, x2), max(x1, x2)
```

chrisb83 replied:
I would prefer to raise an error to point out that something is wrong (could be that the user put 5 and 1 instead of -5 and 1)

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+        if not opt_res.success:
+            raise Exception("solver failed: " + opt_res.message)
+        if np.isnan(opt_res.fun):
+            raise Exception("solver returned nan at x=".format(opt_res.x))
```
These are cryptic Exception messages to process. The user is expecting to receive random variates, instead they get a message about a solver failing. In their minds how will they connect a solver failing with the inability to generate random variates?
If there is an ability for the user to fix a problem then that message should be returned.

chrisb83 replied:
agree, hopefully better now

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+    j, simulated = 1, 0
+    while (j <= maxiter):
+        k = size - simulated
+        # simulate uniform rvs on [x1, x2] and [0,y2]
```
[0, y2]

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+        maxiter = 10 + size * np.log(1 / p_fail) / np.log(1 / (1 - p_accept))
+
+    # start sampling using ratio of uniforms method
+    x = np.zeros(size)
```
Remember that size can be an int, or a tuple of ints. Therefore x could be N-dimensional, not just a vector.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+    x = np.zeros(size)
+    j, simulated = 1, 0
+    while (j <= maxiter):
+        k = size - simulated
```
This won't work if `size = (10, 20, 30)` (for N-Dimensional random variates).

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+        if simulated < size:
+            j += 1
+        else:
+            return x
```
remember x needs to be N-Dimensional, depending on the `size` argument. I would use a vector for `x`, then reshape to the desired `size` when the function returns.

andyfaff commented:
I also forgot to mention that the function should have a [`random_state` keyword argument](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.rvs.html).
This is to ensure a reproducible variate stream.
You can use `scipy._lib._util.check_random_state` for creating a `RandomState` object. You then use this object to seed random numbers from your uniform distributions.

andyfaff commented:
Tests will need to check that:
1) random variates obey e.g. a normal distribution, and also a 'harder' distribution that is the reason for the PR in the first place.
2) check that repeatable random variate streams work.
3) Check that the size argument works for an int and tuple of ints.
4) check that x1, x2 specifications work,
etc.

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+    while (j <= maxiter):
+        k = size - simulated
+        # simulate uniform rvs on [x1, x2] and [0,y2]
+        u1 = distributions.uniform.rvs(loc=x1, scale=(x2 - x1), size=k)
```
why not just use:
```
u1 = np.random.random(size=k) * (x2 - x1) + x1
```
it's going to be closer to the metal. Also instead of `np.random.random` you can use a `RandomState` instance.

e.g.
```
rng = check_random_state(random_state)
...
u1 = rng(size=k) * (x2 - x1) + x1
```

chrisb83 replied:
good point, changed

andyfaff commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5720,179 @@ def rankdata(a, method='average'):
+        k = size - simulated
+        # simulate uniform rvs on [x1, x2] and [0,y2]
+        u1 = distributions.uniform.rvs(loc=x1, scale=(x2 - x1), size=k)
+        v1 = distributions.uniform.rvs(scale=y2, size=k)
```
```
v1 = np.random.random(size=k) * y2
```

chrisb83 commented:
@andyfaff : I revised the code and addressed your comments above, in particular regarding the size parameter. 

I started to add test cases but this is still work in progress

chrisb83 commented:
I did a simple check to see how the method compares against the default rvs method for the generalized inverse Gaussian distribution (https://en.wikipedia.org/wiki/Generalized_inverse_Gaussian_distribution)

Since the CDF is not known in closed form, generating rvs via cdf Inversion is very slow:
generating 1500 rvs takes 119 secs on my computer. Using ratio-of-uniforms, it takes 0.031 secs (x1, x2, y2 are known, see referenced article in the doc string).

Code: 
[test_gen_inverse_gaussian.txt](https://github.com/scipy/scipy/files/1661648/test_gen_inverse_gaussian.txt)

chrisb83 commented:
@andyfaff: I added more  test cases in 32fbb81, I hope this is fairly complete now.

1. i added normal distribution as a test case and the generalized inverse Gauss which is not part of scipy yet and which is a Distribution for which rvs generation is nontrivial (test is a bit ugly since I added the cdf calculation via integrate.quad by hand). I didn't add a test case to check whether finding x1,x2 numerically works since a) it would be rather a test of minimize_scalar b) distributions that come to my mind have generally tails that decay fast and should not make the maximization hard c) there is a disclaimer in the notes saying that one should pay attention when x1, x2 are not specified.

2. "repeatable rv streams" : i tested that results are the same if `size = 9` or `size = (3, 3)` (after reshaping) if the same seed is used. is that what you meant?

3. Tests of size Arguments

4. tested some exception handling of x1, x2, y2

Let me know if you would like to see further tests.

chrisb83 commented:
@andyfaff : Hi, did you already have time to look into the changes? FYI: I started to work on adding the generalized inverse Gaussian to scipy.stats which would rely on this method, see https://github.com/chrisb83/scipy/tree/geninvgauss.

chrisb83 commented:
@andyfaff hi, it would be great if you could provide your feedback on this PR. Thanks

ev-br commented:
@chrisb83 Just wanted to say that this would be a great contribution. Please don't be discouraged by the lack of action here, it's just the bandwidth issue from our side.

chrisb83 commented:
Hi,
just FYI: #8681 that uses `ratio_of_unif` is not finalized yet, but shows that the function can be used to generate rvs as intended. It would be great to get your feedback.

WarrenWeckesser commented:
After digging into numpy's implementation of the hypergeometric distribution sampler, I know a bit more about ratio-of-uniforms method than I did back in January, so I feel more comfortable about reviewing this pull request now.  However, I'll be traveling for a few weeks, so I won't get to it until the end of August.  (If anyone else wants to jump in sooner, please do.)

rgommers commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+        specified (i.e. equal to None), attempt to find the value numerically
+        over the interval from a to b.
+    random_state : None or int or np.random.RandomState instance, optional
+        If int or RandomState, use it for drawing the random variates.
```
Maybe clearer to steal this from the `random_state` docstring in `rv_generic`:
```
        If already a RandomState instance, use it.
        If an int, use a new RandomState instance seeded with seed.
```

rgommers commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+
+    Returns
+    -------
+    rvs : array
```
array --> ndarray

rgommers commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+    Notes
+    -----
+    Given a univariate probability density function (pdf) f and a constant c,
+    define the set A(c) = {(x, y) : 0 < y <= sqrt(f(x/y + c)) }.
```
This needs either double backticks to timeset as code, or a change to ``:math:``.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+    using only the pdf, i.e. no inversion of the cdf is required. Typical
+    choices of c are zero or the mode of f. The set A(c) is a subset of the
+    rectangle R = [-x1, x2] x [0, y2] where
+    - x1 = inf (x - c) sqrt(f(x))
```
I think this needs a blank line above it to render as list. And again double backticks.

rgommers replied:
There are many other formatting issues. Here is the rendered docstring: https://5360-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.stats.rvs_ratio_unif.html#scipy.stats.rvs_ratio_unif (you may need to log in with your GitHub account).

chrisb83 replied:
Where can I find these links?

rgommers commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+
+    Simulate normally distributed random variables. It is easy to compute the
+    bounding rectangle explicitly in that case.
+    >>> f = stats.norm.pdf
```
blank lines around docstring examples needed for them to render correctly

rgommers commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+        if bounds is None:
+            raise Exception("The parameter bounds needs to be specified if " +
+                            var + " is None.")
+        opt_res = optimize.minimize_scalar(g, bounds=bounds, method='Bounded')
```
`Bounded` --> `bounded`? I see we need to fix that in `minimize_scalar`. Method names should be lowercase in such APIs I think. We can still change that before the next release. @pvanmulbregt do you agree?

chrisb83 replied:
I changed it in my code, `minimize_scalar` converts method to lower case anyways.

optimize._minimize.py
``
if callable(method):
        meth = "_custom"
    else:
        meth = method.lower()
``

rgommers commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+                            var + " is None.")
+        opt_res = optimize.minimize_scalar(g, bounds=bounds, method='Bounded')
+        if not opt_res.success:
+            raise Exception("Optimization failed when attempting to find " +
```
You usually never raise an `Exception`, but something more specific. `RuntimeError` is probably a good choice here.

And same above and below.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+    if x1 is None:
+        x1 = find_rect_bounds(f_opt, bounds, "x1")
+    if x2 is None:
+        x2 = -find_rect_bounds(lambda x: -f_opt(x), bounds, "x2")
```
the logic of the two minus signs here deserves a comment

rgommers commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+        # apply rejection method
+        rvs = u1 / v1 + shift
+        accept = (v1**2 <= pdf(rvs))
+        num_accept = sum(accept)
```
style nitpick: better to use `np.sum`

rgommers commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+        if simulated >= N:
+            return np.reshape(x, size1d)
+
+    raise Exception("Only {} of {} random variables".format(simulated, N) +
```
again change to RuntimeError

rgommers commented:
This needs a rebase as well.

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+    The above result (see [1]_, [2]_) can be used to sample random variables
+    using only the pdf, i.e. no inversion of the cdf is required. Typical
+    choices of c are zero or the mode of f. The set A(c) is a subset of the
+    rectangle R = [-x1, x2] x [0, y2] where
```
The interval `[-x1, x2] x [0, y2]` and the ratio of areas below `2 * y2 * (x2 - x1)` don't seem consistent to me. 

I'm looking at the example of a uniform distribution on the interval `[2, 6]` (see the Kinderman/Monahan 1977 paper listed in another comment below), using c=shift=0.  Then `y2=1/sqrt(6-2) = 0.5, x1=1, x2=3`.  The region `A(c)` appears to be the triangle with vertices `(0, 0), (0.5, 1), (0.5, 3)` and area `0.5`,  and trying to reconcile that with the rectangle R, and its implied area `y2 * (x2 - x1)=1`.  Is `R(c)=[-1, 3]x[0, 0.5]` with area 2?  Or is `R(c)=[1, 3]x[0, 0.5]` with area 1?  Certainly the latter doesn't contain `A(c)`, so I'm guessing the former?

pvanmulbregt replied:
The calculation of `x1` in the example above isn’t quite correct. `x1` needs to be calculated over all of the reals, not just the interval [2,6], so x1 should be 0, not 0.5.  A better example might be a uniform distribution on the interval `[-1, 3]`, with `y2=0.5, x1=-0.5 and x2=1.5`.  That suggests the rectangle `R=[x1, x2] x [0, y2]`.

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+
+    Intuitively, the method works well if A(c) fills up most of the
+    enclosing rectangle such that the probability is high that (U, V) lies in
+    A(c) whenever it lies in R(c) as the number of required iterations
```
Is `R(c)` defined in the docstring?

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
+
+    .. [2] W. Hoermann and J. Leydold, "Generating generalized inverse Gaussian
+       random variates", Statistics and Computing, 24(4), p. 547--557, 2014.
+
```
Perhaps mention the 1977 paper which may be the earliest description of the algorithm. It also has useful examples giving the computation of regions for a few distributions.  BiBTex below
@article{Kinderman:1977:CGR:355744.355750,
 author = {Kinderman, A. J. and Monahan, J. F.},
 title = {Computer Generation of Random Variables Using the Ratio of Uniform Deviates},
 journal = {ACM Trans. Math. Softw.},
 issue_date = {Sept. 1977},
 volume = {3},
 number = {3},
 month = sep,
 year = {1977},
 issn = {0098-3500},
 pages = {257--260},
 numpages = {4},
 url = {http://doi.acm.org/10.1145/355744.355750},
 doi = {10.1145/355744.355750},
 acmid = {355750},
 publisher = {ACM},
 address = {New York, NY, USA},
}

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -5939,3 +5942,206 @@ def rankdata(a, method='average'):
+    The above result (see [1]_, [2]_) can be used to sample random variables
+    using only the pdf, i.e. no inversion of the cdf is required. Typical
+    choices of c are zero or the mode of f. The set ``A`` is a subset of the
+    rectangle ``R = [-x1, x2] x [0, y2]`` where
```
Should this be  `R = [x1, x2] x [0, y2]`?  I.e. `x1` instead of `-x1`?

chrisb83 replied:
fixed

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -5939,3 +5942,206 @@ def rankdata(a, method='average'):
+    y2 : float, optional
+        The upper bound of the bounding rectangle in the y-direction.
+        Default is None. In that case, attempt to find bound numerically.
+    shift : float, optional.
```
Is the name `shift` open to being renamed? Would you consider just using `c`?  Reasoning behind the suggestion:
- `c` matches what is used in the `Notes` section
- `c` doesn't have any other meaning in this context which might cause confusion, whereas `shift` has multiple  potential meanings.
- `c` is shorter

chrisb83 replied:
Yes, i am open to renaming it and I agree that `shift` is not self-explanatory, so why not just call it `c` as in the notes.

chrisb83 commented:
I removed the parameter `maxiter` in bec7dc0 for the following reason: 
- to generate a single rv, one has a success probability p given by the ratio of the areas of A and R in each try.
- the first time that success occurs has a geometric distribution (mean 1/p)
- having no success for a single rv until step k is equal to `(1-p)**k`, so even for small p, this will be very unlikely after a while
- if we want to generate N rvs, we do the same experiment N times independently

I think it is enough to display a warning if p is small (<5% in the implementation) because the method will be very slow in that case but there is no need to enforce a limit on the number of iterations . 

I also renamed `shift` to `c`

chrisb83 commented:
@andypfaff do you want to take a look at this PR again? The changes you requests have been implemented.

rgommers commented:
I plan to merge this soonish. @andyfaff @WarrenWeckesser if you still want to review before merge, let me know.

andyfaff commented:
I think the functionality in this PR is a great addition to scipy.

My main reservation about the PR is that the functionality is related to #6466, creating an `rv_continuous` from an arbitrary PDF function. The functionality here is pretty much the `rvs` method that I would expect from such a distribution.
Would it be better to have such an `rv_arbitrary`, with this code supplying the rvs method functionality? Or does it merit standalone inclusion?

rgommers commented:
> Would it be better to have such an `rv_arbitrary`, with this code supplying the rvs method functionality? Or does it merit standalone inclusion?

I think this is a well-known method, has a clean API and solves a real problem (e.g. the generalized inverse Gaussian in gh-8681, pretty hard to do otherwise). It's likely also faster than putting it inside `rv_arbitrary`. So I'd say standalone.

Note that this then doesn't preclude introducing something like `rv_arbitrary` - this is a necessary building block for an `rv_arbitrary.rvs`. Regarding gh-6466, I think that could be useful, but re-reading the discussion there it's mostly a matter of being very clear on use cases it solves and API it offers (plus not too much code duplication with `rv_continuous`). It could use a new summary; the description at the top is outdated (e.g. it talks about histogram, but we now have `rv_histogram`).

WarrenWeckesser commented:
I'm working on it this evening; I'll comment on the PR within the next 24 hours.

WarrenWeckesser commented on scipy/stats/stats.py:
```diff
@@ -5939,3 +5942,195 @@ def rankdata(a, method='average'):
+    is displayed if this ratio is larger than 20.
+
+    If the bounding rectangle is not correctly specified (i.e. if it does not
+    contain ``A``, the algorithm samples from a distribution different from
```
Missing closing parenthesis.

chrisb83 commented:
@WarrenWeckesser thanks for your comments.

Main changes I will do:
- use u, v as in the original paper (since I started to work on this when looking at the GIG paper, I took over the notation), rename to vmin, vmax etc
- move code to a separate file
 
Regarding the option to find bounds: I am fine to remove it, indeed, it makes more sense to find the bounds outside of the function (for speed and to avoid problems to find the min /max). 

> We would certainly use ratio of uniforms, but we'd probably customize the implementation for the specific distribution. 

there are indeed various techniques to speed up the method, but i think it is a nice general tool. for GIG, it is also applied directly for most of the parameter ranges in the paper by Hörmann and Leydold.

rgommers commented:
Splitting `stats.py` into categories sounds like a useful thing to do to me. Not 100% sure we won't run into more import cycle issues though. Maybe split up as needed when we work on it? E.g. now create `_rvs_sampling.py` and put just this function in it; then in a separate PR move the other sampling functions, but not create any other new files.

chrisb83 commented:
I moved the function to a new file `_rvs_sampling.py`. TravisCI passes and the test failure on CircleCI seems unrelated:

> Worker 'gw1' crashed while running 'build/testenv/site-packages/scipy/stats/tests/test_distributions.py::TestLevyStable::()::test_pdf_nolan_samples'


Not sure about AppVeyor though:

> +++++++++++++++++++++++++++++++++++ Timeout 
> [gw1] node down: Not properly terminated
> Replacing crashed worker gw1
> ........................................................................ [ 96%]
> .....
> ........................................................................ [ 99%]
> ================================== FAILURES 
> ___________________ scipy/_lib/tests/test_import_cycles.py ____________________
> [gw1] win32 -- Python 3.7.0 C:\Python37-x64\python.exe
> Worker 'gw1' crashed while running 'scipy/_lib/tests/test_import_cycles.py::test_modules_importable'

chrisb83 commented:
Now all tests pass, so I guess the crash last time was just a coincidence.

@WarrenWeckesser all your comments should be addressed now

chrisb83 commented:
Can this PR be merged? Then i could continue with #8681 at some point. thanks

rgommers commented:
@WarrenWeckesser and @andyfaff looks like all comments from both of you were addressed. Could you check and update your review approval? Would be good to merge this for 1.2.0

andyfaff commented on scipy/stats/_rvs_sampling.py:
```diff
@@ -0,0 +1,156 @@
+
+    Parameters
+    ----------
+    pdf : callable
```
What signature does pdf require?

chrisb83 replied:
it is the pdf without further parameters

andyfaff commented:
I've no further comments at this time. I can't get back to looking at it for a couple of weeks, so I'm happy for it to be merged if others are happy.

WarrenWeckesser commented:
I'll get back to this by Monday, if not sooner.

rgommers commented:
> does anyone recall if we have an established convention for documenting the signature of a callable argument?

Normally we just do:
```
func : callable
    A function with signature, `callable(x, y, ...)` that <etc>
```

chrisb83 commented:
Easy part: I adjusted the documentation.

More difficult: force algorithm to terminate. I removed one approach in bec7dc09110a5fdc53d3da415f731df02e39cefe .
Now i had a better idea: If we want to simulate N rvs, assume we still don't have a single success (i.e. a single (U, V) in the acceptance region A) after i iterations of the loop. this means we have tested whether (U, V) in A i*N times without success. if each event has probability p, this occurs with probability (1-p)**(i*N). Assuming that we only have a success rate of 1/1000, this probability below 2*1e-22 if i*N > 50000. and in that case, i raise an exception. 

BOM: I don't think there is an option to disable it in Spyder and for Windows users, it seems hard to avoid adding it to a file (Notepad adds it according to Wikipedia). Would it really be the only file that has this BOM in scipy? I can check how to remove remove it tomorrow if really needed.

chrisb83 commented:
@WarrenWeckesser BOM is gone as well, very easy with Notepad++, I will try to keep it in mind in the future.

chrisb83 commented:
@WarrenWeckesser it would be great If you take a look before it is too late for 1.2.0. thanks

WarrenWeckesser commented on scipy/stats/_rvs_sampling.py:
```diff
@@ -0,0 +1,169 @@
+    Given a univariate probability density function (pdf) f and a constant c,
+    define the set ``A = {(u, v) : 0 < u <= sqrt(f(v/u + c))}``.
+    If ``(U, V)`` is a random vector uniformly distributed over ``A``,
+    then ``V/U + c`` follows a distribution with pdf ``f``.
```
Be consistent in the markup used for f; currently the text has f, \`f\` and \`\`f\`\`.    I think \`f\` is fine (and use \`c\` for c, too).  The numpy docstring standard says to use single backticks for parameters, and f is not actually a parameter, but it is really just another name for `pdf`, so it might as well be treated the same.  Might as well use single backticks for the isolated occurrences of A and R, too, and just use the double backticks for expressions.  Text in single backticks gets rendered in italics, while double backticks are rendered as monospace text (i.e. code-like), so this isn't perfect, but I think that's a limitation of the reStructuredText that we're stuck with. (Well, it would be possible use :math: markup for all the mathematical variables and expressions, but personally, I view unrendered docstrings more often than I view the web versions, so I am biased against the heavy use of LaTeX markup.)

It might also work fine to remove *all* the backticks around the *mathematical* variables and expressions, and only use those for actual function parameter names and Python code, but I don't know if everyone would agree that that is acceptable.  (It definitely doesn't work if a variable is called `a`, but that is not the case here.) For example,

        define the set A = {(u, v) : 0 < u <= sqrt(f(v/u + c))}.

looks OK to me.

I'd write more, but it might end up as a mini-rant about the limitations of Sphinx and reStructuredText, so I'll stop. :)  The only change that is *really* needed here is to be consistent.

WarrenWeckesser commented on scipy/stats/_rvs_sampling.py:
```diff
@@ -0,0 +1,169 @@
+    the ratio ``area(R) / area(A) = 2 * umax * (vmax - vmin)``, using the fact
+    that the area of ``A`` is equal to 1/2 (Theorem 7.1 in [1]_). A warning
+    is displayed if this ratio is larger than 20. Moreover, if the sampling
+    fails to generate a single random variate after 50'000 iterations (i.e.
```
50000 (no apostrophe)

chrisb83 commented:
@WarrenWeckesser I addressed the main points you mentioned. I mainly used single backticks now, with the exception of longer expressions . rgommers asked for double backticks in his review  some time ago and I also prefer it for readability. I think this is not consistent in scipy, maybe we have a separate discussion on that after 1.2.0.

rgommers commented:
> I mainly used single backticks now, with the exception of longer expressions . rgommers asked for double backticks in his review some time ago and I also prefer it for readability. I think this is not consistent in scipy, maybe we have a separate discussion on that after 1.2.0.

It isn't consistent everywhere, but what it should be in every case is reasonably well-defined. Basic rule: references to modules/functions/classes in our API as well as references to function parameters should be single backticks; everything else that's code-like needs double backticks for typesetting as code.

The current state of this PR looks good enough to me.

andyfaff commented:
Could we have that basic rule written down to refer to, doc formatting is
something that is always tripping me up.

On Sun., 4 Nov. 2018, 12:19 Ralf Gommers <notifications@github.com wrote:

> I mainly used single backticks now, with the exception of longer
> expressions . rgommers asked for double backticks in his review some time
> ago and I also prefer it for readability. I think this is not consistent in
> scipy, maybe we have a separate discussion on that after 1.2.0.
>
> It isn't consistent everywhere, but what it should be in every case is
> reasonably well-defined. Basic rule: references to
> modules/functions/classes in our API as well as references to function
> parameters should be single backticks; everything else that's code-like
> needs double backticks for typesetting as code.
>
> The current state of this PR looks good enough to me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/8293#issuecomment-435633899>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAq51p6jgCvvXW7zmi54lmRqmEbp6Vw7ks5urkCFgaJpZM4RdSnz>
> .
>

ilayn commented on scipy/stats/_rvs_sampling.py:
```diff
@@ -0,0 +1,169 @@
+
+    exp_iter = 2 * (vmax - vmin) * umax  # rejection constant (see [1])
+    if exp_iter > 20:
+        msg = "The expected number of iterations to generate a single random" \
```
Example: 

```python
long_string = ('Lorem ipsum dolor sit amet, consectetur adipisicing elit, '
               'sed do eiusmod tempor incididunt ut labore et dolore magna '
               'aliqua. Ut enim ad minim veniam, quis nostrud exercitation.')
```

ilayn commented on scipy/stats/_rvs_sampling.py:
```diff
@@ -0,0 +1,169 @@
+        Defining number of random variates (default is 1).
+    c : float, optional.
+        Shift parameter of ratio-of-uniforms method, see Notes. Default is 0.
+    random_state : None or int or np.random.RandomState instance, optional
```
None is implied by the signature. 

```python
    random_state : int, numpy.random.RandomState, optional
```

is sufficient.

rgommers commented:
> Could we have that basic rule written down to refer to, doc formatting is something that is always tripping me up.

I see it's not very explicit at https://numpydoc.readthedocs.io/en/latest/format.html, although what's mentioned there about backticks is correct. If we improve that doc is that enough, or would you expect to find this somewhere else?

chrisb83 commented:
@WarrenWeckesser i updated the PR yesterday and addressed your comments. Please review again.

WarrenWeckesser commented:
Looks good.  My only remaining request is to squash the commits before merging.  (And yeah, I should have squashed the commits over at https://github.com/scipy/scipy/pull/9435.  I didn't think to check the commits there until after hitting the merge button.)

rgommers commented:
Everyone happy now, so squash-merged. Great to have this one in! Thanks @chrisb83 and all reviewers!
